service/s3/s3manager/batch.go

```diff
@@ -0,0 +1,225 @@
+package s3manager
+
+import (
+	"io"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+)
+
+// BatchDeleteIterator is an interface that uses the scanner pattern to
+// iterate through what needs to be deleted.
+type BatchDeleteIterator interface {
+	Next() bool
+	Err() error
+	DeleteObject() BatchDeleteObject
+}
+
+// BatchDelete will use the s3 package's service client to perform a batch
+// delete.
+type BatchDelete struct {
+	Client    s3iface.S3API
+	BatchSize int
+}
+
+// NewBatchDelete will return a new delete client that can delete a batched amount of
+// objects.
+//
+// Example:
+//	batcher := s3manager.NewBatchDelete(client, size)
+//
+//	objects := []BatchDeleteObject{
+//		{
+//			Object:	&s3.DeleteObjectInput {
+//				Key: aws.String("key"),
+//				Bucket: aws.String("bucket"),
+//			},
+//		},
+//	}
+//
+//	iter := s3manager.NewBatchDeleteIterator([]BatchDeleteObject{})
+//	if err := batcher.Delete(iter); err != nil {
+//		return err
+//	}
+func NewBatchDelete(client s3iface.S3API, size int) *BatchDelete {
+	return &BatchDelete{
+		Client:    client,
+		BatchSize: size,
+	}
+}
+
+func next(index, size int, err error) bool {
+	return index < size && err == nil
+}
+
+// BatchDeleteObject is a wrapper object for calling the batch delete operation.
+type BatchDeleteObject struct {
+	Object *s3.DeleteObjectInput
+}
+
+type batchDeleteIter struct {
+	Objects []BatchDeleteObject
+	index   int
+	err     error
+}
+
+func (batcher *batchDeleteIter) Next() bool {
+	return next(batcher.index, len(batcher.Objects), batcher.err)
+}
+
+func (batcher *batchDeleteIter) Err() error {
+	return batcher.err
+}
+
+func (batcher *batchDeleteIter) DeleteObject() BatchDeleteObject {
+	object := batcher.Objects[batcher.index]
+	batcher.index++
```

Comment:
`Next` should be the function to increment the index. `DeleteObject` should just return the value at the current index.